File: 07/day07.go
package main

import (
    "strings"
    "github.com/deosjr/adventofcode2020/lib"
)

func recFind(containedBy map[string][]string, color string) map[string]struct{} {
    m := map[string]struct{}{}
    containers := containedBy[color]
    for _, c := range containers {
        subm := recFind(containedBy, c)
        m[c] = struct{}{}
        for k,v := range subm {
            m[k] = v
        }
    }
    return m
}

type container struct {
    num int64
    color string
}

func recCount(contains map[string][]container, color string) int64 {
    var sum int64
    containers := contains[color]
    for _, c := range containers {
        sum += c.num
        sum += c.num * recCount(contains, c.color)
    }
    return sum
}

func main() {
    contains := map[string][]container{}
    containedBy := map[string][]string{}
    readfunc := func(line string) {
        firstsplit := strings.Split(strings.TrimSpace(line), " bags contain ")
        color := firstsplit[0]
        content := firstsplit[1]
        if content == "no other bags." {
            return
        }
        split := strings.Split(content, ", ")
        for _, c := range split {
            trimmed := strings.TrimRight(c, "bags.")
            trimmed = strings.TrimSpace(trimmed)
            num := lib.MustParseInt(string(trimmed[0]))
            col := trimmed[2:]
            container := container{num:num, color:col}
            contains[color] = append(contains[color], container)
            containedBy[col] = append(containedBy[col], color)
        }
    }
    lib.ReadFileByLine(7, readfunc)

    p1 := recFind(containedBy, "shiny gold")
    lib.WritePart1("%d", len(p1))

    p2 := recCount(contains, "shiny gold")
    lib.WritePart2("%d", p2)
}

File: 07/day07.pl
:- ['../lib/io.pl'].

:- dynamic contains/3.

set_contains(Color, Set, New) :-
    findall(X, contains(X, _, Color), List),
    maplist([X,Y]>>(set_contains(X, [], Y)), List, Sets),
    foldl(union, [List|Sets], Set, New).

part1(Ans) :- 
    set_contains("shiny gold", [], Set),
    length(Set, Ans).

count_contains(Color, Sum) :-
    findall(N-X, contains(Color, N, X), List),
    maplist([N-X,Y]>>(count_contains(X, Count), Y=N-Count), List, Sets),
    foldl([N-X,Y,Z]>>(Z #= Y + N + N*X), Sets, 0, Sum).

part2(Ans) :-
    count_contains("shiny gold", Ans).

parse --> parse_line, "\n", parse.
parse --> parse_line, blanks, eos.

parse_line --> parse_color(C), " bags contain ", parse_contents(C).

parse_contents(_) --> "no other bags.". 
parse_contents(C) --> parse_content(N, S), ".", {assertz(contains(C,N,S))}.
parse_contents(C) --> parse_content(N, S), ", ", parse_contents(C), {assertz(contains(C,N,S))}.

parse_content(N, S) --> integer(N), " ", parse_color(S), " ", ("bag"; "bags").

parse_color(S) --> string_without(" ", Mod), " ", string_without(" ", Col), 
    {string_codes(MStr, Mod), string_codes(CStr, [32|Col]), string_concat(MStr, CStr, S)}.

run :-
    input_stream(7, parse),
    part1(Ans1),
    write_part1(Ans1),
    part2(Ans2),
    write_part2(Ans2).
